Add sinatra instrumentation adapter  #67

duonoid opened the pull request:
# Overview
Adds sinatra instrumentation adapter.

## Details
* Implements tracing using sinatra middleware and extension
* Adds runnable example for demonstration and testing

## Related
* Similar: #148
* Effort toward resolving #67

fbogsany commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/extensions/tracer_extension.rb:
```diff
@@ -0,0 +1,37 @@
+              def render(engine, data, *)
+                Sinatra::Adapter.tracer.in_span(
+                  'sinatra.render_template',
+                  kind: :server,
```
`kind:` shouldn't be specified here. It will default to `:internal`. This span isn't a server span - it's parent span is (the middleware span).

duonoid replied:
fixed :heavy_check_mark: by removing `kind:`

fbogsany commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/extensions/tracer_extension.rb:
```diff
@@ -0,0 +1,37 @@
+                Sinatra::Adapter.tracer.in_span(
+                  'sinatra.render_template',
+                  kind: :server,
+                  with_parent: Sinatra::Adapter.tracer.current_span
```
This shouldn't be necessary. The `current_span` will be used as the parent by default.

duonoid replied:
I've removed `with_parent:` :heavy_check_mark:

fbogsany commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/extensions/tracer_extension.rb:
```diff
@@ -0,0 +1,37 @@
+                ) do |span|
+                  template_name = data.is_a?(Symbol) ? data : :literal
+                  span.set_attribute('sinatra.template_name', template_name.to_s)
+                  output = super
```
Do you need to assign to `output` or can you just call `super` here?

duonoid replied:
I've addressed some other unused variables as well.  :heavy_check_mark:

fbogsany commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/middlewares/tracer_middleware.rb:
```diff
@@ -0,0 +1,57 @@
+          end
+
+          def trace_request(span, env)
+            span.set_attribute('http.url', env['PATH_INFO'])
```
Can this be passed as part of the `attributes:` hash to `in_span`?

duonoid replied:
I've moved the call to `attributes:`, which I'm assuming is performance-related (due to the mutex in the sdk implementation of `set_attributes`?). :heavy_check_mark:

fbogsany commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/middlewares/tracer_middleware.rb:
```diff
@@ -0,0 +1,57 @@
+          def trace_response(span, env, resp)
+            status, _headers, _response_body = resp
+
+            span.set_attribute('http.status_code', status)
```
We should set the `span.status` here as well.

duonoid replied:
Hopefully on the right track here... I had to add `Trace::Status.from_http_status(code) => Status`, which didn't _exactly_ map 1-to-1 http -> gRPC status.

duonoid commented:
This PR also uses #154.

duonoid commented:
I could extract the "gRPC status" changes to a separate PR, if needed.

mwear commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/middlewares/tracer_middleware.rb:
```diff
@@ -0,0 +1,53 @@
+              kind: :server,
+              with_parent_context: parent_context(env)
+            ) do |span|
+              app.call(env).tap { |resp| trace_response(span, env, resp) }
```
dd-trace-rb has the ability to opt-in to collecting additional headers: https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/contrib/faraday/middleware.rb#L45-L46.

Should we keep this functionality?

fbogsany replied:
Didn't see the opt-in on the lines you pointed at, but the [semantic conventions for HTTP clients](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#http-client) include the option of `peer.hostname` and `peer.port` or `peer.ip` and `peer.port`, so I think it's worth considering their inclusion here. That said, the semantic conventions favour `http.url`:

> One of the following sets of attributes is required (in order of usual preference unless for a particular web client/framework it is known that some other set is preferable for some reason; all strings must be non-empty):
>
> * http.url
> * http.scheme, http.host, http.target
> * http.scheme, peer.hostname, peer.port, http.target
> * http.scheme, peer.ip, peer.port, http.target

fbogsany replied:
I think you mean https://github.com/DataDog/dd-trace-rb/blob/ab135f65c4d761ddbac02d4d0f04e3c17382218f/lib/ddtrace/contrib/sinatra/tracer_middleware.rb#L30-L32

I'd lean toward "no" for the initial implementation, just to keep it simple and focussed on the current semantic specifications.

mwear replied:
lol... yes... that's the correct link

mwear commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/middlewares/tracer_middleware.rb:
```diff
@@ -0,0 +1,53 @@
+
+          def trace_response(span, env, resp)
+            status, _headers, _response_body = resp
+
```
It similarly has the ability to opt-in to collecting response_headers: https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/contrib/sinatra/tracer_middleware.rb#L37-L39.

Should we preserve this behavior?

fbogsany commented on api/lib/opentelemetry/trace/status.rb:
```diff
@@ -19,9 +19,56 @@ class Status
+      # @param code Numeric HTTP status
+      #
+      # @return Status
+      def self.from_http_status(code) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength
```
This has been extracted to #157 and merged, so can be removed from here.

duonoid commented:
I have sync'd changes per recent discussions on #148 (API changes, tests, etc.)

mwear commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/adapter.rb:
```diff
@@ -0,0 +1,67 @@
+
+        private
+
+        def register_tracer_extension
```
We are likely going to need an `InstrumentationRegistry` that is part of the OpenTelemetry::API to ensure that instrumentation is installed only once. Is the point of this code to fake that we have a registry so that it can be easily changed later?

If so, I'm wondering if it wouldn't be better for us to just have a class level `installed?` flag for now, and update it to use the actual registry later?

duonoid replied:
I've switched to the `installed?` flag.

fbogsany commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra.rb:
```diff
@@ -0,0 +1,30 @@
+
+      # Convenience method to access the nested module name
+      def name
+        Module.nesting[0].to_s
```
TIL

fbogsany commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/adapter.rb:
```diff
@@ -0,0 +1,63 @@
+
+          private
+
+          def tracer_factory
```
This method seems unnecessary. It is only called in one place (`install`) and is *not* used in one other place it could be (`tracer`). I don't think it provides enough value.

duonoid replied:
removed :heavy_check_mark:

fbogsany commented on adapters/sinatra/example/server.rb:
```diff
@@ -0,0 +1,43 @@
+  )
+)
+
+OpenTelemetry::Adapters::Sinatra.install(name: 'sinatra-example', version: '1.0')
```
The `config` hash isn't used anywhere. It seems confusing to specify `name` and `version` here when they're not used in anyway and don't actually mean anything.

duonoid replied:
Updated to avoid passing `name` and `version` :heavy_check_mark:

fbogsany commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/adapter.rb:
```diff
@@ -0,0 +1,63 @@
+
+          register_tracer_extension
+
+          :installed
```
I'm probably missing a bunch of earlier discussion. What is the intent behind the `:installed` and `:already_installed` return values? I don't see the return value used in the example.

fbogsany replied:
Discussed IRL: this is mostly documentation & makes testing more convenient.

fbogsany commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/extensions/tracer_extension.rb:
```diff
@@ -0,0 +1,36 @@
+                  'sinatra.render_template'
+                ) do |span|
+                  template_name = data.is_a?(Symbol) ? data : :literal
+                  span.set_attribute('sinatra.template_name', template_name.to_s)
```
`template_name` can be computed before the `span` is created. It will be more efficient to provide the attribute to `in_span`. I.e.:
```ruby
def render(_engine, data, *)
  template_name = data.is_a?(Symbol) ? data : :literal
  Sinatra::Adapter.tracer.in_span(
    'sinatra.render_template',
    attributes: {'sinatra.template_name' => template_name.to_s}
  ) do
    super
  end
end
```

duonoid replied:
Updated to pass data when span is created (for performance). :heavy_check_mark:

benedictfischer09 commented on adapters/sinatra/lib/opentelemetry/adapters/sinatra/middlewares/tracer_middleware.rb:
```diff
@@ -0,0 +1,53 @@
+          attr_reader :app
+
+          def parent_context(env)
+            OpenTelemetry::Adapters::Sinatra::Adapter.propagator.extract(env)
```
Is it the job of a propagator to inspect this payload and decide whether it is trustworthy? Can someone craft headers malicious headers that would cause confusion in tracing systems?  Maybe a simple flag to always start a new trace instead of accepting existing context.  Or, this might be a case where offering hooks into the instrumentation would enable more sophisticated behavior like client code looking at the request and deciding whether to allow the context it contains to be extracted

duonoid commented:
I have addressed feedback to this point, as well as added some fixes/workarounds for a build issue regarding circleci + bundler-2.1.x (including 2.1.4) + appraisal (2.2.0).

mwear commented:
@duonoid Could you take a look at the failures on Ruby 2.7? It's same story as https://github.com/open-telemetry/opentelemetry-ruby/pull/148#issuecomment-572670786.

duonoid commented:
@mwear I have applied the same solution (remove appraisal from circleci, use `rake test`) as the one that fixed https://github.com/open-telemetry/opentelemetry-ruby/pull/148#issuecomment-572670786, as described here: https://github.com/open-telemetry/opentelemetry-ruby/pull/148#issuecomment-572692800

duonoid commented:
@mwear I've updated https://github.com/open-telemetry/opentelemetry-ruby/pull/148#issuecomment-572781259

mwear commented:
Would you mind resolving these conflicts and I'll get this merged!

duonoid commented:
conflicts: resolved :heavy_check_mark:

mwear commented:
Thanks @duonoid! 🚀
